core/primitives/src/sr25519.rs

```diff
@@ -0,0 +1,330 @@
+// Copyright 2017-2018 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+// tag::description[]
+//! Simple sr25519 (Schnorr-Ristretto) API.
+// end::description[]
+
+use base58::{FromBase58, ToBase58};
+use blake2_rfc;
+use rand::rngs::OsRng;
+use schnorrkel::{signing_context, Keypair, MiniSecretKey, PublicKey};
+use sha2::Sha512;
+
+#[cfg(feature = "std")]
+use serde::{de, Deserialize, Deserializer, Serializer};
+
+// signing context
+const SIGNING_CTX: &'static [u8] = b"polkadot transaction";
```

Comment:
that doesn't seem right, since we're in the substrate code.

Also, the way that schnorrkel is meant to be used is that signatures for each purpose will have their own signing-context (to create domain separation, prevent length-extension attacks).

Using it with a single global context lowers the security of the library.

Reply:
We should eventually be able to write roughly this whenever `const fn` gets stabilized and becomes popular enough for merlin to use:
```
const POLKADOT_PAYMENT_CTX: SigningContext = signing_context(b"PolkadotPayment");
```
but right now this would require a lazy_static I think.

It's entirely possible that I overthought this interface though ;)  In fact, one could add convenience methods like
```
fn sign_bytes(&self, context: &'static [u8], bytes: &[u8]) {
    self.sign(signing_context(contect).bytes(bytes))
}
```
but there are places in polkadot where you want the `hash256` builder method instead of the `bytes` builder method because `Sha2` or `Shake128` sound faster than merlin's Keccek variant.  And these builder help keep the interface simple in this case. 

Also, there are cases where you sign very small things, like certificates and the VRF in block production, where you could build the merlin transcript manually, maybe using an extension trait like merlin recommends.

Reply:
`b"substrate transaction"` or even `b"transaction"` is fine.

Reply:
Anyways @rphmeier is correct that using the same context for everything breaks the domain separation that signing_context helps provide.  We could do domain seperation in other ways like by substrate and polkadot having their own versions of the signing_context function that do "substrate" or "polkadot" before whatever context the users provides.  We should however do the domain separation though because the alternative is adding more key types, which wastes space on chain.

Reply:
Based on what I am reading so far, adding a set of `sign() / verify()` methods that accept the context is a good way to go. API-wise it also looks clean. we will end up with two groups of exposed functions: 
- default context: `pub fn verify_strong(sig, message, pubkey) -> bool` / `pub fn sign(message)`
- custom context: `pub fn verify_strong_ctx(sig, message, pubkey) -> bool` / `pub fn sign_ctx(message)`

as [already suggested](https://github.com/paritytech/substrate/pull/1706#discussion_r254360043) by @burdges

Reply:
It's fine as it is. Polkadot and Substrate domain separate are *as I just said* managed at a place further down in the stack where it makes more sense for that data to be integrated into the signing data.

If there's any use at all to this context parameter it's to distinguish between using the key for Substrate-family chain signatures and the key being used in other non-Substrate family contexts (which are probably nil, but then I'm unconvinced of the utility of exposing context at this low-level).